glfs.c

```diff
@@ -748,6 +750,59 @@ static int tcmu_glfs_discard(struct tcmu_device *dev, struct tcmulib_cmd *cmd,
 	return TCMU_STS_NO_RESOURCE;
 }
 
+static int tcmu_glfs_writesame(struct tcmu_device *dev,
+			       struct tcmulib_cmd *cmd,
+			       uint64_t offset, uint64_t length,
+			       struct iovec *iov, size_t iov_cnt)
+{
+	struct glfs_state *state = tcmu_get_dev_private(dev);
+	glfs_cbk_cookie *cookie;
+	ssize_t ret;
+
+	cookie = calloc(1, sizeof(*cookie));
+	if (!cookie) {
+		tcmu_dev_err(dev, "Could not allocate cookie: %m\n");
+		goto out;
+	}
+	cookie->dev = dev;
+	cookie->cmd = cmd;
+	cookie->length = 0;
+	cookie->op = TCMU_GLFS_WRITESAME;
+
+	ret = glfs_zerofill_async(state->gfd, offset, length, glfs_async_cbk, cookie);
```

Comment:
Does the glfs function assume it is always zeros in the buffer? The ceph WS command has a check for if the buffer to write is all zeros or non zero'd values. For the latter it will drop down to doing normal writes. If your zerofill command does not have a similar check then I think here you want to add a check and if non zeroed data then return TCMU_STS_NOT_HANDLED so the common code will handle it.

Reply:
Currently the glfs function do not receive any buffered data from here, just the offset and length, then fill it as zero, that's all.

Yeah, this make sense here. Very handsome advice.
Thanks Mike, will address this here.